<a href="https://colab.research.google.com/github/hc0rd31r0/Bootcamp_Data_Science/blob/main/projeto-final/projeto_final_tratamento_dados.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src="https://github.com/hc0rd31r0/Bootcamp_Data_Science/blob/main/projeto-final/img/Banner_Bootcamp.png?raw=true">

#**Projeto Final de Conclusão de Curso**
###Bootcamp Data Science Aplicada 2 by [Alura](https://www.alura.com.br/) 
####Autor: Helton Cordeiro
e-mail: heltoncordeiro@gmail.com

Junho-Agosto/2021.


---

# Objetivo desse notebook

#### Tratar os dados disponibilizados pelo Hospital Sírio-Libanês no site [Kaggle](https://www.kaggle.com/S%C3%ADrio-Libanes/covid19) para o projeto final de conclusão do curso Bootcamp de Data Science da Alura.

#### Ao final teremos uma planilha que será utilizada para processamento dos modelos de Machine Learning (ML) que estão em um notebook próprio.


O dataset contém informações anonimizadas e agrupadas em:

* Informação demográfica - 3 variáveis
* Doenças pré-existentes - 9 variáveis
* Resultados de exames de sangue - 36 variáveis
* Sinais vitais - 6 variáveis



No total são 54 variáveis que foram expandidos quando pertinentes à média (mean), mediana (median), máximo (max), mínimo (min), diferença (diff) e diferença relativa (diff_rel), sendo:

1. diff = max - min
2. diff_rel = diff / median



---



## Importação dos dados

Está sendo utilizada uma base de dados obtida no dia 06/07/2021 e está na pasta dados do projeto final no GitHub.

In [1]:
import pandas as pd
import numpy as np
import funcoes

In [2]:
#para recarregar as funções em caso de alteração sem precisar reiniciar o Colab.
#from importlib import reload
#reload(funcoes)

In [3]:
from funcoes import preenche_tabela, remove_corr_var

In [4]:
help(preenche_tabela)

Help on function preenche_tabela in module funcoes:

preenche_tabela(dados)
    Função para preenher os valores nulos do dataframe desconsiderando as informações 
    das janelas de atendimento que o paciente foi internado (ICU = 1)
    
    Parâmetros
    ----------
      dados: dataFrame.
    
    Retorno
    -------
      dfretorno: dataFrame com o resultado do processamento.



In [5]:
help(remove_corr_var)

Help on function remove_corr_var in module funcoes:

remove_corr_var(dados, valor_corte=0.95)
    Função para remover as colunas do dataframe com valor de correlação.
    
    Parâmetros
    ----------
      dados: dataFrame.
      valor_corte: Valor de corte da correlação, default=0.95
    
    Retorno
    -------
      dfretorno: dataFrame com o resultado do processamento.



In [6]:
dados_raw = pd.read_excel("https://github.com/hc0rd31r0/Bootcamp_Data_Science/blob/main/projeto-final/dados/Kaggle_Sirio_Libanes_ICU_Prediction.xlsx?raw=true")

### Informações da base de dados

In [7]:
print(40*"--x")
print("Nosso dataframe tem (%s) linhas e (%s) colunas" %(dados_raw.shape[0], dados_raw.shape[1]))
print("Sendo 5 linhas por paciente, totalizando (%s) pacientes" %(int(dados_raw.shape[0] / 5)))
print(40*"--x")
display(dados_raw.head(7))

--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x
Nosso dataframe tem (1925) linhas e (231) colunas
Sendo 5 linhas por paciente, totalizando (385) pacientes
--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x--x


PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                         0            1  ...       0-2    0
1                         0            1  ...       2-4    0
2                         0            1  ...       4-6    0
3                         0            1  ...      6-12    0
4                         0            1  ...  ABOVE_12    1
5                         1            1  ...       0-2    1
6                         1            1  ...       2-4    1

[7 rows x 231 columns]

In [8]:
print("ICU:", dados_raw['ICU'].unique())
print("WINDOW:", dados_raw['WINDOW'].unique())
print("AGE_PERCENTIL:", np.sort(dados_raw['AGE_PERCENTIL'].unique()))
print("Total de valores nulos (nan):", dados_raw.isna().sum().sum())
display(dados_raw.describe())

ICU: [0 1]
WINDOW: ['0-2' '2-4' '4-6' '6-12' 'ABOVE_12']
AGE_PERCENTIL: ['10th' '20th' '30th' '40th' '50th' '60th' '70th' '80th' '90th'
 'Above 90th']
Total de valores nulos (nan): 223863


PATIENT_VISIT_IDENTIFIER  ...          ICU
count               1925.000000  ...  1925.000000
mean                 192.000000  ...     0.267532
std                  111.168431  ...     0.442787
min                    0.000000  ...     0.000000
25%                   96.000000  ...     0.000000
50%                  192.000000  ...     0.000000
75%                  288.000000  ...     1.000000
max                  384.000000  ...     1.000000

[8 rows x 229 columns]

# Análise Exploratória

* A coluna ```PATIENT_VISIT_IDENTIFIER``` é o identificado do paciente.

* ```AGE_ABOVE65``` indica se o paciente tem idade acima de 65 anos.

* ```AGE_PERCENTIL``` representa a idade em décadas. ex. um paciente com 55 anos é apresentado como 50th.

* ```GENDER``` identifica o gênero do paciente, mas por questões de anonimização não podemos afirmar se 0 (zero) é Masculino e 1 (um) é Feminino ou vice-versa.

* ```WINDOW``` apresenta os seguintes valores ```['0-2', '2-4', '4-6', '6-12', 'ABOVE_12']``` que representam as horas de atendimento de um paciente entre 0 e 2 horas, de 2 a 4 horas, de 4 a 6 horas, de 6 a 12 horas e acima de 12 horas de atendimento.

* ```ICU``` é o que desejamos prever, se 1 = Internou ou 0 = não internou.

* As colunas de ```DISEASE GROUPING 1``` até ```OTHER``` identificam doenças pré-existentes do paciente.

* As demais colunas são resultados de exames de sangue (36 variáveis em 180 colunas) e de Sinais vitais (6 variáveis em 36 colunas)

* Identificamos muitas colunas com valores nulos que precisam ser tratadas para serem utilizadas pelos modelos de Machine Learning.

# Utilização dos dados das janelas de atendimento

De acordo com a descrição do problema e com a orientação do Hospital Sírio Libânes, temos diversas janelas de atendimento e seguem abaixo as orientações ee como devem ser utilizá-las.

5 janelas de atendimento e:

* **Exemplo 1**

![](https://www.googleapis.com/download/storage/v1/b/kaggle-user-content/o/inbox%2F1591620%2Fb1bc424df771a4d2d3b3088606d083e6%2FTimeline%20Example%20Best.png?generation=1594740856017996&alt=media)


* **Exemplo 2**

![](https://www.googleapis.com/download/storage/v1/b/kaggle-user-content/o/inbox%2F1591620%2F77ca2b4635bc4dd7800e1c777fed9de1%2FTimeline%20Example%20No.png?generation=1594740873237462&alt=media)


***Devemos atentar para a regra de não utilzar os dados da janela de atendimento quando o paciente é INTERNADO.***

# Tratamento dos dados

## UTI na 1ª Janela
Conforme explicação acima, os dados não podem ser utilizados após o paciente ser admitido na UTI, assim os registros de 32 pacientes devem ser descartados pois o internamento aconteceu na 1ª janela de atendimento, entre 0-2 horas.

In [9]:
a_remover = dados_raw.query('WINDOW == "0-2" and ICU == 1')['PATIENT_VISIT_IDENTIFIER'].values
a_remover

array([  1,  17,  27,  41,  47,  48,  54,  62,  85,  90, 104, 106, 109,
       117, 144, 176, 202, 206, 226, 233, 239, 261, 270, 282, 286, 306,
       315, 329, 359, 362, 366, 379])

In [10]:
dados = dados_raw.query("PATIENT_VISIT_IDENTIFIER not in @a_remover")
print(dados.shape)

(1765, 231)


## Preenchendo os dados faltantes

Utilizando os métodos fillna(method="ffill") e fillna(method="bfill") para preencher os valores com os resultados das linhas anterior e seguinte.

As features (colunas) foram separadas pelo pessoal da Hospital Sírio-Libanês.
* Da coluna 13, até a final -2  são informações das medições.
* As colunas de 1 a 13 categorizam nosso paciente.

De acordo com as explicações acima, os dados não devem ser utilizar após a admissão do paciente na UTI, assim vamos filtrar as informações quando o ICU != 1.

A função **```preenche_tabela```** está no arquivo funcoes.py.

In [11]:
dados = preenche_tabela(dados)
dados.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                         0            1  ...       0-2    0
1                         0            1  ...       2-4    0
2                         0            1  ...       4-6    0
3                         0            1  ...      6-12    0
4                         0            1  ...  ABOVE_12    1

[5 rows x 231 columns]

## Removendo os nulos remanescentes

In [12]:
dados = dados.dropna()
dados.shape

(1485, 231)

In [13]:
len(dados['PATIENT_VISIT_IDENTIFIER'].unique())

351

## 1 (um) registro por paciente

Vamos utilizar a 1ª janela de atendimento, pois quanto antes identificarmos a necessidade de um leito de UTI melhor.

In [14]:
def prepare_window(rows):
    if(np.any(rows["ICU"])):
        rows.loc[rows["WINDOW"]=="0-2", "ICU"] = 1
    return rows.loc[rows["WINDOW"] == "0-2"]

In [15]:
dados_limpos = dados.groupby("PATIENT_VISIT_IDENTIFIER").apply(prepare_window).reset_index(drop=True)
dados_limpos.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ... WINDOW  ICU
0                         0            1  ...    0-2    1
1                         2            0  ...    0-2    1
2                         3            0  ...    0-2    0
3                         4            0  ...    0-2    0
4                         5            0  ...    0-2    0

[5 rows x 231 columns]

In [16]:
dados_limpos.shape

(294, 231)

# Mais procedimentos

Com os dados preenchidos, os nulos remanescentes removidos, os pacientes atendidos na 1ª janela descartados e 1 registro por paciente com as informações importantes, vamos finalizar a nossa planilha de dados removendo colunas desnecessárias e categorizando as possíveis.

### Categorizando a coluna ```AGE_PERCENTIL```

Na coluna AGE_PERCENTIL temos a idade do paciente mas separado em décadas e em número ordinal, ex.: se o paciente em 53 anos, no dataset é apresentado como 50th e num campo string.


In [17]:
dados_limpos['AGE_PERCENTIL'] = dados_limpos['AGE_PERCENTIL'].str.replace('Above 9','10').str.extract(r'(.+?)th')
dados_limpos = dados_limpos.astype({'AGE_PERCENTIL': 'int'})
dados_limpos['AGE_PERCENTIL'] = dados_limpos['AGE_PERCENTIL'] / 10

### Removendo as colunas ```WINDOW``` e ```PATIENT_VISIT_IDENTIFIER```

A coluna WINDOW perdeu relevância quando optamos por escolher a 1ª janela de atendimento e a coluna PATIENT_VISIT_IDENTIFIER possui um identificador número único por paciente o que é irrelevante para nossos modelos de ML.

In [18]:
dados_limpos.drop(['PATIENT_VISIT_IDENTIFIER', 'WINDOW'], axis=1, inplace=True)
dados_limpos.head()

AGE_ABOVE65  AGE_PERCENTIL  ...  OXYGEN_SATURATION_DIFF_REL  ICU
0            1            6.0  ...                   -1.000000    1
1            0            1.0  ...                   -0.961262    1
2            0            4.0  ...                   -1.000000    0
3            0            1.0  ...                   -0.980333    0
4            0            1.0  ...                   -0.980129    0

[5 rows x 229 columns]

### Removendo as colunas altamente correlacionadas

Nossa base de dados ainda tem muitas colunas (labels) (229) e isso pode facilitar o  overfit e afetar nossos modelos, mas quais labels podemos remover?

Vamos identificar as colunas com alta correlação e para isso vamos gerar uma matriz de correlação, mas antes vamos remover as colunas que não nos interessam para a correlação, que são:
1. AGE_ABOVE65 = Acima de 65 anos. 
2. AGE_PERCENTIL = idade por décadas.
3. GENDER	= Gênero. 
4. ICU = nosso y, o que queremos prever

vamos desconsiderar da matriz de correlação com

> ```dados_limpos.iloc[:,3:-1]```

In [19]:
dados_limpos.iloc[:,3:-1].corr().abs()

DISEASE GROUPING 1  ...  OXYGEN_SATURATION_DIFF_REL
DISEASE GROUPING 1                         1.000000  ...                    0.046261
DISEASE GROUPING 2                         0.177343  ...                    0.121114
DISEASE GROUPING 3                         0.316506  ...                    0.020760
DISEASE GROUPING 4                         0.035122  ...                    0.033468
DISEASE GROUPING 5                         0.317468  ...                    0.106645
...                                             ...  ...                         ...
BLOODPRESSURE_SISTOLIC_DIFF_REL            0.024205  ...                    0.508116
HEART_RATE_DIFF_REL                        0.149202  ...                    0.394567
RESPIRATORY_RATE_DIFF_REL                  0.059516  ...                    0.387704
TEMPERATURE_DIFF_REL                       0.023078  ...                    0.463901
OXYGEN_SATURATION_DIFF_REL                 0.046261  ...                    1.000000

[225 rows x 225 columns]

Para realizarmos a operação corretamente vamos precisar de alguns comandos do Pandas e da Numpy

A função **```remove_corr_var```** está no arquivo funcoes.py.


In [20]:
dados_limpos_sem_corr = remove_corr_var(dados_limpos, 0.95)
dados_limpos_sem_corr

AGE_ABOVE65  AGE_PERCENTIL  ...  HEART_RATE_DIFF_REL  ICU
0              1            6.0  ...            -1.000000    1
1              0            1.0  ...            -0.747001    1
2              0            4.0  ...            -1.000000    0
3              0            1.0  ...            -0.956805    0
4              0            1.0  ...            -0.986481    0
..           ...            ...  ...                  ...  ...
289            0            4.0  ...            -1.000000    1
290            1           10.0  ...            -1.000000    0
291            0            5.0  ...            -1.000000    1
292            0            4.0  ...            -1.000000    0
293            0            5.0  ...            -1.000000    0

[294 rows x 95 columns]

# Salvando o resultado final



No final de todo o processo de obtenção, limpeza e tratamento dos dados acabamos com um dataframe com:

In [21]:
print("(%s) linhas e (%s) colunas." %(dados_limpos_sem_corr.shape[0], dados_limpos_sem_corr.shape[1]))

(294) linhas e (95) colunas.


In [22]:
dados_limpos.to_excel("/content/drive/MyDrive/databases/Kaggle_Sirio_Libanes_ICU_Prediction-tratado.xls")
dados_limpos_sem_corr.to_excel("/content/drive/MyDrive/databases/Kaggle_Sirio_Libanes_ICU_Prediction-tratado-sem-corr.xls")

## Referências

* [Hospital Sírio-Libanês](https://www.hospitalsiriolibanes.org.br/Paginas/nova-home.aspx)
* [Kaggle - COVID-19 - Clinical Data to assess diagnosis - Sírio Libanês](https://www.kaggle.com/S%C3%ADrio-Libanes/covid19)
* [Pandas](https://pandas.pydata.org/pandas-docs/stable/index.html)
* [Pandas.corr()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.corr.html)
* [Pandas.Where()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.where.html)
* [Numpy.triu()](https://numpy.org/doc/stable/reference/generated/numpy.triu.html) (Upper triangle of an array.)
* [Sobre Correlação](https://pt.wikipedia.org/wiki/Correla%C3%A7%C3%A3o)
